### PR TITLE
go: extra functionality to help Trillian

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: go/client/ctclient/ctclient.go [hunk 1]
```diff
@@ -64,8 +66,27 @@ func addChain(ctx context.Context, logClient *client.LogClient) {
...
+	}
+
+	// Examine the leaf to see if it looks like a pre-certificate.
+	isPrecert := false
+	leaf, err := x509.ParseCertificate(chain[0].Data)
+	if err == nil {
+		count, _ := x509util.OidInExtensions(x509.OidExtensionCTPoison, leaf.Extensions)
+		if count > 0 {
+			isPrecert = true
+			fmt.Print("Uploading pre-certificate to log\n")
```
Comment (AlCutter) [inline go/client/ctclient/ctclient.go]: log.Info?

File: go/client/ctclient/ctclient.go [hunk 2]
```diff
@@ -81,27 +102,74 @@ func getRoots(ctx context.Context, logClient *client.LogClient) {
...
+		log.Fatal("No -last option supplied")
+	}
+	entries, err := logClient.GetEntries(ctx, *getFirst, *getLast)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, entry := range entries {
+		ts := entry.Leaf.TimestampedEntry
+		when := ctTimestampToTime(ts.Timestamp)
+		fmt.Printf("Index=%d Timestamp=%v ", entry.Index, when)
```
Comment (AlCutter) [inline go/client/ctclient/ctclient.go]: log.Infof? (or log.V maybe?)
Or build up a string to log later on if you want it all on one line?
(and elsewhere)

File: go/client/ctclient/ctclient.go [hunk 3]
```diff
@@ -81,27 +102,74 @@ func getRoots(ctx context.Context, logClient *client.LogClient) {
...
+			log.Fatalf("Unhandled log entry type %d", entry.Leaf.TimestampedEntry.EntryType)
 		}
 	}
 }
+
+func showCert(cert ct.ASN1Cert) {
+	if *textOut {
+		c, err := x509.ParseCertificate(cert.Data)
+		if err != nil {
+			log.Printf("Error parsing certificate: %q", err.Error())
```
Comment (AlCutter) [inline go/client/ctclient/ctclient.go]: Errorf? (and elsewhere)

File: go/x509/x509.go [hunk 4]
```diff
@@ -759,6 +759,38 @@ func (h UnhandledCriticalExtension) Error() string {
...
+// RemoveCTPoison takes a DER-encoded TBSCertificate and removes the CT poison extension,
+// and returns the result, still as a DER-encoded TBSCertificate.  This function will
+// fail if there is not exactly 1 CT poison extension present.
+func RemoveCTPoison(tbsData []byte) ([]byte, error) {
+	var tbs tbsCertificate
+	rest, err := asn1.Unmarshal(tbsData, &tbs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse TBSCertificate: %v", err)
+	} else if len(rest) > 0 {
+		return nil, fmt.Errorf("trailing data after TBSCertificate: %v", err)
```
Comment (AlCutter) [inline go/x509/x509.go]:
```go
else if rLen = len(ret); rLen > 0 { 
   return nil, fmt.ErrorF("...%d", rLen)
}
```

File: go/x509/x509.go [hunk 5]
```diff
@@ -1344,6 +1376,8 @@ var (
 	oidExtensionNameConstraints       = []int{2, 5, 29, 30}
 	oidExtensionCRLDistributionPoints = []int{2, 5, 29, 31}
 	oidExtensionAuthorityInfoAccess   = []int{1, 3, 6, 1, 5, 5, 7, 1, 1}
+	// RFC 6962 s3.1
+	OidExtensionCTPoison = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3}
```
Comment (AlCutter) [inline go/x509/x509.go]: g/Oid/s//OID/g

Comment (daviddrysdale) [inline go/x509/x509.go]: Done.

File: go/client/ctclient/ctclient.go (hunk 3 above)
Comment (daviddrysdale) [inline go/client/ctclient/ctclient.go]: That's only in `glog`, not in base [`log`](https://golang.org/pkg/log/).

File: go/client/ctclient/ctclient.go (hunk 1 above)
Comment (daviddrysdale) [inline go/client/ctclient/ctclient.go]: At the moment all of the normal output from `ctclient` is just `fmt.Printf` -- all of the `log`s are for error conditions.

File: go/client/ctclient/ctclient.go (hunk 3 above)
Comment (AlCutter) [inline go/client/ctclient/ctclient.go]: D'oh.

File: go/client/ctclient/ctclient.go (hunk 1 above)
Comment (AlCutter) [inline go/client/ctclient/ctclient.go]: oke doke.

**Pull request merged**